### PR TITLE
Feature/5820 dataframe extend groupby

### LINKED ISSUE #5820: Extend GroupBy API of DataFrame with ability to iterate groups and rows
### System information

- **OS version/distro**: Any
- **.NET Version (eg., dotnet --info)**: Any

### Issue

- **What did you do?**

Used GroupBy method of dataframe

- **What happened?**

Got GroupBy object. GroupBy object doesn't provide any easy way to iterate the groups and rows inside the group

- **What did you expect?**

I expected ability to have API similar to GroupBy operation in LINQ with ability to iterate groups and DataFrameRows inside th group (having IEnumerable<IGrouping<out TKey, out TElement>> interface)

### PR DESCRIPTION
**Pull request by asmirnov82**
Fixes #5820

Extend DataFrame GroupBy operation by adding new property Groupings. This property returns collection of IGrouping objects (the same way as LINQ GroupBy operation does). This allows to use syntacs like this:

`
var groups = dataFrame.GroupBy<TKey>(columnName).Groupings.ToDictionary(g => g.Key, g => g.ToList());
foreach (DataFrameRow row  in groups[KeyValue])
{ 
      // any code to work with row object
}
`

### CODE REVIEW

**Comment by eerhardt** [comment]
> using System;

Need a copyright header on all new files.

---
Refers to: test/Microsoft.Data.Analysis.Tests/DataFrameGroupByTests.cs:1 in 0580710. [](commit_id = 05807104b471661aba732af88599372423f7bdb5, deletion_comment = False)

File: test/Microsoft.Data.Analysis.Tests/DataFrameGroupByTests.cs [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+using Xunit;
+
+namespace Microsoft.Data.Analysis.Tests
+{
+    public class DataFrameGroupByTests
+    {
+        [Fact]
+        public void TestGroupingWithTKeyTypeofString()
+        {
+            int lenght = 11;
```
Comment (eerhardt) [inline test/Microsoft.Data.Analysis.Tests/DataFrameGroupByTests.cs:14]:
```suggestion
            int length = 11;
```

File: src/Microsoft.Data.Analysis/Grouping.cs [hunk 2]
```diff
@@ -0,0 +1,32 @@
+﻿using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+
+namespace Microsoft.Data.Analysis
+{
+    public class Grouping<TKey> : IGrouping<TKey, DataFrameRow>
```
Comment (eerhardt) [inline src/Microsoft.Data.Analysis/Grouping.cs:9]: Can you add some xml documentation to all new public API?

File: test/Microsoft.Data.Analysis.Tests/DataFrameGroupByTests.cs (hunk 1 above)
Comment (eerhardt) [reply test/Microsoft.Data.Analysis.Tests/DataFrameGroupByTests.cs:14]: Repeated everywhere.

File: src/Microsoft.Data.Analysis/GroupBy.cs [hunk 3]
```diff
@@ -70,7 +72,7 @@ public abstract class GroupBy
         public abstract DataFrame Mean(params string[] columnNames);
     }
 
-    public class GroupBy<TKey> : GroupBy
+    public class GroupBy<TKey> : GroupBy
```
Comment (eerhardt) [inline src/Microsoft.Data.Analysis/GroupBy.cs:75]: (nit) this whitespace change should be reverted

File: src/Microsoft.Data.Analysis/GroupBy.cs [hunk 4]
```diff
@@ -464,5 +466,12 @@ public override DataFrame Mean(params string[] columnNames)
             return ret;
         }
 
+        public IEnumerable<IGrouping<TKey, DataFrameRow>> Groupings
```
Comment (eerhardt) [inline src/Microsoft.Data.Analysis/GroupBy.cs:469]: Do we need a new `Groupings` property? Or could we just implement `IEnumerable<IGrouping<TKey, DataFrameRow>>` on `GroupBy<TKey>` itself?

Comment (pgovind) [reply src/Microsoft.Data.Analysis/GroupBy.cs:469]: I think this should be implemented directly on the `GroupBy<TKey>` type too. Especially since this PR is now exposing a `GroupBy<TKey>` method

File: test/Microsoft.Data.Analysis.Tests/DataFrameGroupByTests.cs [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+
+            //Create test dataframe (numbers starting from 0 up to lenght)
+            DataFrame df = MakeTestDataFrame(lenght);
+
+            //Use wrong type for grouping
+            Assert.Throws<InvalidCastException>(() => df.GroupBy<double>("Tens"));
+        }
+
+
+        private DataFrame MakeTestDataFrame(int length)
```
Comment (pgovind) [inline test/Microsoft.Data.Analysis.Tests/DataFrameGroupByTests.cs:96]: While you are here, would you mind renaming this to `MakeTestDataFrameWithParityAndTensColumns`? Just so it's clear what columns are being created.

File: src/Microsoft.Data.Analysis/GroupBy.cs [hunk 6]
```diff
@@ -464,5 +466,12 @@ public override DataFrame Mean(params string[] columnNames)
...
+        {
+            get 
+            {
+                return _keyToRowIndicesMap.Select(kvp => new Grouping<TKey>(kvp.Key, kvp.Value.Select(index => _dataFrame.Rows[index]).ToArray()));
```
Comment (pgovind) [inline src/Microsoft.Data.Analysis/GroupBy.cs:473]: The `ToArray` here will make the computation here greedy right? Would it be better to keep this lazy (since the return type here is `IEnumerable<IGrouping<TKey, DataFrameRow>>`)?  I think just removing the `ToArray` will work since DataFrameRow implements `IEnumerable` (I may be slightly wrong here. I didn't try to compile this)

File: src/Microsoft.Data.Analysis/GroupBy.cs (hunk 4 above)
Comment (asmirnov82) [reply src/Microsoft.Data.Analysis/GroupBy.cs:469]: @eerhardt, I agree with you, I also prefer not to have extra property. But there is a small problem here:  GroupBy class has it's own implementation of Count(), First() and other methods, that are provided by LINQ extensions to IEnumerable interface. Moreover all these methods have different return type. For example Count() returns dataframe and not Int. First()  also returns dataframe and not IGrouping. In this case, using of new LINQ method will require explıcıt casting:

`var grouping = df.GroupBy<string>("Parity");
DataFrame count  = grouping.Count(); // GroupBy class own method
int count  = ((IEnumerable<IGrouping<string, DataFrameRow>>)grouping).Count(); //LINQ method
`
Do you have any ideas how this can be solved?

File: src/Microsoft.Data.Analysis/Grouping.cs (hunk 2 above)
Comment (asmirnov82) [reply src/Microsoft.Data.Analysis/Grouping.cs:9]: As GroupBy exposes only IGrouping interface and not Grouping<TKey> class, I changed it to be an internal class inside GroupBy<TKey>. I decided it's better, that increase the amount of public entities, if it is not really required

File: test/Microsoft.Data.Analysis.Tests/DataFrameGroupByTests.cs (hunk 1 above)
Comment (asmirnov82) [reply test/Microsoft.Data.Analysis.Tests/DataFrameGroupByTests.cs]: Not sure, that I got correctly what is required here. I splitted test where length was changed into 2 (segregated test for corner cases into extra one). Other tests don't depend on lenght of the created Dataframe and it can be used as a parameter, for example changeing the code like this:

```
[Theory]
[InlineData(5)]
[InlineData(12)]
public void TestPrimitiveColumnSort(int lenght) {...}
```

File: src/Microsoft.Data.Analysis/GroupBy.cs (hunk 3 above)
Comment (asmirnov82) [reply src/Microsoft.Data.Analysis/GroupBy.cs]: fixed

File: test/Microsoft.Data.Analysis.Tests/DataFrameGroupByTests.cs (hunk 5 above)
Comment (asmirnov82) [reply test/Microsoft.Data.Analysis.Tests/DataFrameGroupByTests.cs]: fixed

File: src/Microsoft.Data.Analysis/GroupBy.cs (hunk 6 above)
Comment (asmirnov82) [reply src/Microsoft.Data.Analysis/GroupBy.cs]: fixed

File: test/Microsoft.Data.Analysis.Tests/DataFrameGroupByTests.cs (hunk 1 above)
Comment (eerhardt) [reply test/Microsoft.Data.Analysis.Tests/DataFrameGroupByTests.cs]: > Not sure, that I got correctly what is required here

Did you see the suggested change above? It is correcting the spelling of the word `length` from the type-o `lenght`.

File: src/Microsoft.Data.Analysis/Grouping.cs (hunk 2 above)
Comment (eerhardt) [reply src/Microsoft.Data.Analysis/Grouping.cs]: In the current form, it doesn't even need to be `internal`. It can just be `private`.

File: src/Microsoft.Data.Analysis/GroupBy.cs [hunk 7]
```diff
@@ -464,5 +493,12 @@ public override DataFrame Mean(params string[] columnNames)
             return ret;
         }
 
+        public IEnumerable<IGrouping<TKey, DataFrameRow>> Groupings
```
Comment (eerhardt) [inline src/Microsoft.Data.Analysis/GroupBy.cs:496]: New public API needs XML documentation.

File: test/Microsoft.Data.Analysis.Tests/DataFrameGroupByTests.cs (hunk 1 above)
Comment (asmirnov82) [reply test/Microsoft.Data.Analysis.Tests/DataFrameGroupByTests.cs]: Got it now. Fixed

File: src/Microsoft.Data.Analysis/Grouping.cs (hunk 2 above)
Comment (asmirnov82) [reply src/Microsoft.Data.Analysis/Grouping.cs]: Fixed

File: src/Microsoft.Data.Analysis/GroupBy.cs (hunk 7 above)
Comment (asmirnov82) [reply src/Microsoft.Data.Analysis/GroupBy.cs:499]: Done

File: src/Microsoft.Data.Analysis/DataFrame.cs [hunk 8]
```diff
@@ -367,6 +367,28 @@ public GroupBy GroupBy(string columnName)
...
+                
+        /// <summary>
+        /// Groups the rows of the <see cref="DataFrame"/> by unique values in the <paramref name="columnName"/> column.
+        /// </summary>
+        /// <typeparam name="TKey">Type of column used for grouping</typeparam>
+        /// <param name="columnName">The column used to group unique values</param>
+        /// <returns>A GroupBy object that stores the group information.</returns>
+        public GroupBy<TKey> GroupBy<TKey>(string columnName)
+        {
+            int columnIndex = _columnCollection.IndexOf(columnName);
```
Comment (eerhardt) [inline src/Microsoft.Data.Analysis/DataFrame.cs:379]: Can this call into the existing `GroupBy GroupBy(string columnName)` method above and then do the appropriate cast? That way we don't need to duplicate this code.

File: src/Microsoft.Data.Analysis/Grouping.cs [hunk 9]
```diff
@@ -0,0 +1,32 @@
+﻿using System;
```
Comment (pgovind) [inline src/Microsoft.Data.Analysis/Grouping.cs]: You'll need to add the license text up here. Just copying it from another file in the repo will work fine.

Comment (pgovind) [reply src/Microsoft.Data.Analysis/Grouping.cs]: Never mind, github was showing me an old view

File: src/Microsoft.Data.Analysis/GroupBy.cs [hunk 10]
```diff
@@ -464,5 +493,15 @@ public override DataFrame Mean(params string[] columnNames)
             return ret;
         }
 
+        /// <summary>
+        /// Returns the collection of Grouping objects, where each object represent as set fo DataFrameRows having the same Key
```
Comment (pgovind) [inline src/Microsoft.Data.Analysis/GroupBy.cs:497]:
```suggestion
        /// Returns a collection of Grouping objects, where each object represent a set of DataFrameRows having the same Key
```

File: src/Microsoft.Data.Analysis/GroupBy.cs (hunk 4 above)
Comment (pgovind) [reply src/Microsoft.Data.Analysis/GroupBy.cs:499]: Ah, yea that is a problem. The `Count, First and other` methods are convenient APIs to inspect the `GroupBy` quickly. For now, I think we can create this new property. In the future, we can implement `IEnumerable<IGrouping<TKey, DataFrameRow>>` on `GroupBy<TKey>` and refactor the `Count` and `First` implementations to possibly just use equivalent  `Linq` methods instead. 

I wonder if we should name this property `Groups` instead of `Groupings`? @eerhardt

Comment (eerhardt) [reply src/Microsoft.Data.Analysis/GroupBy.cs:499]: > I wonder if we should name this property Groups instead of Groupings?

I don't think I have a strong opinion either way... Maybe since the interface is named `IGrouping`, calling this `Groupings` seems appropriate.

**Comment by pgovind** [comment]
I pushed a commit to fix the remaining comments, so we can be ready for the next preview release. Great work here @asmirnov82, thanks!

**Review by pgovind — APPROVED** [review]

**Review by eerhardt — APPROVED** [review]
:shipit:
